### PR TITLE
Implemented tag system

### PR DESCRIPTION
**Pull request by n-Ultima**
The tag system, and everything else implemented, follows the conventional C# analyzers and common formatting.

Changes include:

* Created the Tag Model

* Implemented the Tag Model as a DbSet

* Created the services relating to Tags in DataAccessLayer

* Implemented the service via commands in GeneralModule

### CODE REVIEW

**Review by n-Ultima — COMMENTED** [review]
Edited RequireTagAuthorization attribute to read from the appsettings.json file.

File: CWE.Data/Context/CWEDbContext.cs [hunk 1]
```diff
@@ -26,5 +26,10 @@ public CWEDbContext(DbContextOptions options)
         /// Gets or sets the <see cref="DbSet{TEntity}"/> containing all requests.
         /// </summary>
         public DbSet<Request> Requests { get; set; }
+
+        /// <summary>
+        /// Gets or sets the the <see cref="DbSet{TEntity}"/> containg all tags.
```
Comment (hendrikdemir) [inline CWE.Data/Context/CWEDbContext.cs]:
```suggestion
        /// Gets or sets the <see cref="DbSet{TEntity}"/> containing all tags.
```
Fixed the sentence

File: CWE.Data/DataAccessLayer.cs [hunk 2]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
             return await this.dbContext.Requests
                 .ToListAsync();
         }
+
+        /// <summary>
+        /// Checks if the tag name provided exists in the database.
+        /// </summary>
+        /// <param name="tagName">The name of the tag to fetch..</param>
+        /// <returns>A <see cref="Tag"/> depending on if the tag exists in the database.</returns>
+        public async Task<Tag> FetchTagAsync(string tagName)
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs]:
```suggestion
        public async Task<Tag> GetTag(string tagName)
```
Same naming as previous methods

File: CWE.Data/DataAccessLayer.cs [hunk 3]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
             return await this.dbContext.Requests
                 .ToListAsync();
         }
+
+        /// <summary>
+        /// Checks if the tag name provided exists in the database.
+        /// </summary>
+        /// <param name="tagName">The name of the tag to fetch..</param>
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs]:
```suggestion
        /// <param name="tagName">The name of the tag to get.</param>
```

File: CWE.Data/DataAccessLayer.cs [hunk 4]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+        /// <returns>A <see cref="Tag"/> depending on if the tag exists in the database.</returns>
+        public async Task<Tag> FetchTagAsync(string tagName)
+        {
+            var tag = await this.dbContext.Tags
+                .Where(x => x.Name == tagName)
+                .SingleOrDefaultAsync();
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs]:
```suggestion
                .FirstOrDefaultAsync();
```
Use same LINQ queries like previous methods.

File: CWE.Data/DataAccessLayer.cs [hunk 5]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+
+            if (tag == null)
+            {
+                throw new ArgumentException("The name provided does not match an existing tag in the database.");
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs:151]: Should never throw an exception, either return some result (with an object, i.e. NotFound) or null (see previous methods).

File: CWE.Data/DataAccessLayer.cs [hunk 6]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+            return tag;
+        }
+
+        /// <summary>
+        /// Creates a tag.
+        /// </summary>
+        /// <param name="name">The name that the tag should have.</param>
+        /// <param name="ownerId">The owner's user ID of the tag.</param>
+        /// <param name="content">The content that the tag holds.</param>
+        /// <returns><see cref="void"/>.</returns>
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs:173]:
```suggestion
        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
```

File: CWE.Data/DataAccessLayer.cs [hunk 7]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+        public async Task CreateTagAsync(string name, ulong ownerId, string content)
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs]:
```suggestion
        public async Task CreateTag(string name, ulong ownerId, string content)
```
Same naming as other methods.

File: CWE.Data/DataAccessLayer.cs [hunk 8]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+        {
+            var tag = await this.dbContext.Tags
+                .Where(x => x.Name == name)
+                .AnyAsync();
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs:178]:
```suggestion
                .FirstOrDefaultAsync(x => x.Name == name);
```
Same LINQ queries as other methods.

File: CWE.Data/DataAccessLayer.cs [hunk 9]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+            if (tag)
+            {
+                throw new Exception("A tag with this name already exists.");
+            }
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs:182]:
```suggestion

            if (tag)
            {
                throw new Exception("A tag with this name already exists.");
            }
```
Whitespace before if-statement.

File: CWE.Data/DataAccessLayer.cs [hunk 10]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+        /// <param name="content">The content that the tag holds.</param>
+        /// <returns><see cref="void"/>.</returns>
+        public async Task CreateTagAsync(string name, ulong ownerId, string content)
+        {
+            var tag = await this.dbContext.Tags
+                .Where(x => x.Name == name)
+                .AnyAsync();
+            if (tag)
+            {
+                throw new Exception("A tag with this name already exists.");
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs:181]: Should not throw an exception.

File: CWE.Data/DataAccessLayer.cs [hunk 11]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+            }
+
+            this.dbContext.Tags.Add(new Tag
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs:184]:
```suggestion
            this.dbContext.Add(new Tag
```
No need to specify DbSet on context, see previous methods.

File: CWE.Data/DataAccessLayer.cs [hunk 12]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+                Content = content,
+            });
+            await this.dbContext.SaveChangesAsync();
+        }
+
+        /// <summary>
+        /// Deletes a tag from the database.
+        /// </summary>
+        /// <param name="tagName">The name of the tag to delete.</param>
+        /// <returns><see cref="void"/>.</returns>
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs:197]:
```suggestion
        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
```
Proper return summary.

File: CWE.Data/DataAccessLayer.cs [hunk 13]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+        public async Task DeleteTagAsync(string tagName)
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs]:
```suggestion
        public async Task DeleteTag(string tagName)
```

File: CWE.Data/DataAccessLayer.cs [hunk 14]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+        {
+            var tag = await this.dbContext.Tags
+                .Where(x => x.Name == tagName)
+                .SingleOrDefaultAsync();
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs]:
```suggestion
                .FirstOrDefaultAsync(x => x.Name == tagName);
```

File: CWE.Data/DataAccessLayer.cs [hunk 15]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+
+            if (tag == null)
+            {
+                throw new ArgumentException("The tag provided was not found in the database.");
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs:206]: Should not throw an exception, see other comments.

File: CWE.Data/DataAccessLayer.cs [hunk 16]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+            }
+
+            this.dbContext.Tags.Remove(tag);
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs:209]:
```suggestion
            this.dbContext.Remove(tag);
```
No need to refer to Tags DbSet.

File: CWE.Data/DataAccessLayer.cs [hunk 17]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+            await this.dbContext.SaveChangesAsync();
+        }
+
+        /// <summary>
+        /// Edits the content of a tag, and modifies it to the new content.
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs:214]:
```suggestion
        /// Modifies the content of the tag.
```

File: CWE.Data/DataAccessLayer.cs [hunk 18]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+                .SingleOrDefaultAsync();
+            if (tag == null)
+            {
+                throw new ArgumentException("The tag provided does not match any currently in the database.");
+            }
+
+            if (tag.OwnerId != executorId)
+            {
+                throw new InvalidOperationException("You cannot modify a tag that you do not own.");
+            }
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs]:
```suggestion
        /// <param name="newContent">The content that the new tag should hold.</param>
        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
        public async Task EditTagContent(string tagName, string newContent)
        {
            var tag = await this.dbContext.Tags
                .FirstOrDefaultAsync(x => x.Name == tagName);
                
            if (tag == null)
            {
                throw new ArgumentException("The tag provided does not match any currently in the database.");
            }
```
Proper return summary, different method name and the owner check should be done prior to running this. This is the data access layer, it is only responsible for managing data, not checking if all conditions (whether or not the edit is allowed) are being met.

File: CWE.Data/DataAccessLayer.cs [hunk 19]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+            await this.dbContext.SaveChangesAsync();
+        }
+
+        /// <summary>
+        /// Transfers the ownership of the current tag to a new specified owner.
+        /// </summary>
+        /// <param name="tagName">The name of the tag to transfer.</param>
+        /// <param name="executorId">The person attempting to transfer ownership of the tag.</param>
+        /// <param name="newOwnerId">The new owner of the tag's ID.</param>
+        /// <returns><see cref="void"/>.</returns>
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs:245]:
```suggestion
        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
```
Proper return summary.

File: CWE.Data/DataAccessLayer.cs [hunk 20]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+            }
+
+            if (tag.OwnerId != executorId)
+            {
+                throw new InvalidOperationException("You do not own this tag, so I cannot transfer ownership.");
+            }
+
+            tag.OwnerId = newOwnerId;
+            await this.dbContext.SaveChangesAsync();
+        }
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs]:
```suggestion
        public async Task EditTagOwner(string tagName, ulong executorId, ulong newOwnerId)
        {
            var tag = await this.dbContext.Tags
                .FirstOrDefaultAsync(x => x.Name == tagName)
                
            if (tag == null)
            {
                throw new ArgumentException("The tag provided does not match any inside the database.");
            }

            tag.OwnerId = newOwnerId;
            await this.dbContext.SaveChangesAsync();
        }
```
Again, DAL is not responsible for checking various preconditions.

File: CWE.Data/DataAccessLayer.cs [hunk 21]
```diff
@@ -132,5 +134,122 @@ public async Task<IEnumerable<Request>> GetRequests()
...
+        /// <param name="newOwnerId">The new owner of the tag's ID.</param>
+        /// <returns><see cref="void"/>.</returns>
+        public async Task TransferTagOwnershipAsync(string tagName, ulong executorId, ulong newOwnerId)
+        {
+            var tag = await this.dbContext.Tags
+                .Where(x => x.Name == tagName)
+                .SingleOrDefaultAsync();
+            if (tag == null)
+            {
+                throw new ArgumentException("The tag provided does not match any inside the database.");
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs:253]: Should not throw an exception.

File: CWE/Common/RequireTagAuthoriazation.cs [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+    using System.Text;
+    using System.Threading.Tasks;
+    using CWE.Data;
+    using Discord.Commands;
+    using Discord.WebSocket;
+    using Microsoft.Extensions.Configuration;
+    using Microsoft.Extensions.DependencyInjection;
+
+    /// <summary>
+    /// Attribute used to check if the user is able to create tags.
```
Comment (hendrikdemir) [inline CWE/Common/RequireTagAuthoriazation.cs]:
```suggestion
    /// Attribute used to check if the user is part of the regulars or associates.
```

File: CWE/Common/RequireTagAuthoriazation.cs [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+    /// </summary>
+    public class RequireTagAuthoriazation : PreconditionAttribute
```
Comment (hendrikdemir) [inline CWE/Common/RequireTagAuthoriazation.cs]:
```suggestion
    public class RequirePromoted : PreconditionAttribute
```

File: CWE/Modules/General.cs [hunk 24]
```diff
@@ -234,5 +234,63 @@ public async Task CancelCampaign(SocketGuildUser user)
                 await this.Context.Channel.SendMessageAsync(embed: error);
             }
         }
+
+        /// <summary>
+        /// Command used to call a tag.
+        /// </summary>
+        /// <param name="tagName">The tag to execute.</param>
+        /// <returns>The execution of a tag, <see cref="Task"/>.</returns>
```
Comment (hendrikdemir) [inline CWE/Modules/General.cs:243]:
```suggestion
        /// <returns>A <see cref="Task"/> representing the result of the asynchronous operation.</returns>
```

File: CWE/Modules/General.cs [hunk 25]
```diff
@@ -234,5 +234,63 @@ public async Task CancelCampaign(SocketGuildUser user)
...
+        [Command("tag")]
+        public async Task ExecuteTag(string tagName)
```
Comment (hendrikdemir) [inline CWE/Modules/General.cs:245]:
```suggestion
        public async Task Tag(string tagName)
```
Simplified name.

File: CWE/Modules/General.cs [hunk 26]
```diff
@@ -234,5 +234,63 @@ public async Task CancelCampaign(SocketGuildUser user)
...
+        {
+            var tag = await this.DataAccessLayer.FetchTagAsync(tagName);
```
Comment (hendrikdemir) [inline CWE/Modules/General.cs]:
```suggestion
            var tag = await this.DataAccessLayer.GetTag(tagName);
```
New name.

File: CWE/Modules/General.cs [hunk 27]
```diff
@@ -234,5 +234,63 @@ public async Task CancelCampaign(SocketGuildUser user)
...
+            await this.ReplyAsync(tag.Content);
+        }
+
+        /// <summary>
+        /// Command used to create a tag.
+        /// </summary>
+        /// <param name="tagName">The tag's name.</param>
+        /// <param name="content">The content that the tag should hold.</param>
+        /// <returns>The creation of a tag, <see cref="Task"/>.</returns>
```
Comment (hendrikdemir) [inline CWE/Modules/General.cs:256]:
```suggestion
        /// <returns>A <see cref="Task"/> representing the result of the asynchronous operation.</returns>
```

File: CWE/Modules/General.cs [hunk 28]
```diff
@@ -234,5 +234,63 @@ public async Task CancelCampaign(SocketGuildUser user)
...
+        [Command("tag create")]
+        [RequireTagAuthoriazation]
```
Comment (hendrikdemir) [inline CWE/Modules/General.cs]:
```suggestion
        [RequirePromoted]
```

File: CWE/Modules/General.cs [hunk 29]
```diff
@@ -234,5 +234,63 @@ public async Task CancelCampaign(SocketGuildUser user)
...
+        public async Task CreateTag(string tagName, [Remainder] string content)
+        {
+            var tag = await this.DataAccessLayer.FetchTagAsync(tagName);
```
Comment (hendrikdemir) [inline CWE/Modules/General.cs]:
```suggestion
            var tag = await this.DataAccessLayer.GetTag(tagName);
```

File: CWE/Modules/General.cs [hunk 30]
```diff
@@ -234,5 +234,63 @@ public async Task CancelCampaign(SocketGuildUser user)
...
+            if (tag != null)
+            {
+                throw new ArgumentException("The tag provided already exists, so I can't create one with the matching name.");
```
Comment (hendrikdemir) [inline CWE/Modules/General.cs:264]: Should not throw an exception but return error.

File: CWE/Modules/General.cs [hunk 31]
```diff
@@ -234,5 +234,63 @@ public async Task CancelCampaign(SocketGuildUser user)
...
+            }
+
+            await this.DataAccessLayer.CreateTagAsync(tagName, this.Context.User.Id, content);
```
Comment (hendrikdemir) [inline CWE/Modules/General.cs]:
```suggestion
            await this.DataAccessLayer.CreateTag(tagName, this.Context.User.Id, content);
```

File: CWE/Modules/General.cs [hunk 32]
```diff
@@ -234,5 +234,63 @@ public async Task CancelCampaign(SocketGuildUser user)
...
+        }
+
+        /// <summary>
+        /// Command used to transfer ownership of a tag.
+        /// </summary>
+        /// <param name="tagName">The name of the tag to transfer.</param>
+        /// <param name="newOwnerId">The new owner's ID value.</param>
+        /// <returns>The ownership action of transering a tag. <see cref="Task"/>.</returns>
```
Comment (hendrikdemir) [inline CWE/Modules/General.cs:275]:
```suggestion
        /// <returns>A <see cref="Task"/> representing the result of the asynchronous operation.</returns>
```

File: CWE/Modules/General.cs [hunk 33]
```diff
@@ -234,5 +234,63 @@ public async Task CancelCampaign(SocketGuildUser user)
...
+        [Command("tag transfer")]
+        [RequireTagAuthoriazation]
+        public async Task TranserTag(string tagName, ulong newOwnerId)
+        {
+            // The method already handles verification, so no need to check here.
+            await this.DataAccessLayer.TransferTagOwnershipAsync(tagName, this.Context.User.Id, newOwnerId);
```
Comment (hendrikdemir) [inline CWE/Modules/General.cs]:
```suggestion
            await this.DataAccessLayer.EditTagOwner(tagName, this.Context.User.Id, newOwnerId);
```

Comment (hendrikdemir) [reply CWE/Modules/General.cs]: Should add checks and return i.e. confirmation of the change.

File: CWE/Modules/General.cs [hunk 34]
```diff
@@ -234,5 +234,63 @@ public async Task CancelCampaign(SocketGuildUser user)
...
+        }
+
+        /// <summary>
+        /// The command used to edit a currently existing tag's response.
+        /// </summary>
+        /// <param name="tagName">The tag to edit.</param>
+        /// <param name="newContent">The content that should be put in place of the old content.</param>
+        /// <returns>Transfership of a tag<see cref="Task"/>.</returns>
```
Comment (hendrikdemir) [inline CWE/Modules/General.cs]:
```suggestion
        /// <returns>A <see cref="Task"/> representing the result of the asynchronous operation.</returns>
```

File: CWE/Modules/General.cs [hunk 35]
```diff
@@ -234,5 +234,63 @@ public async Task CancelCampaign(SocketGuildUser user)
...
+        [Command("tag edit")]
+        [RequireTagAuthoriazation]
```
Comment (hendrikdemir) [inline CWE/Modules/General.cs]:
```suggestion
        [RequirePromoted]
```

File: CWE/Modules/General.cs [hunk 36]
```diff
@@ -234,5 +234,63 @@ public async Task CancelCampaign(SocketGuildUser user)
...
+        public async Task EditTag(string tagName, [Remainder] string newContent)
+        {
+            await this.DataAccessLayer.EditTagContentAsync(tagName, this.Context.User.Id, newContent);
```
Comment (hendrikdemir) [inline CWE/Modules/General.cs]: Should add checks if user is allowed to edit tag plus send i.e. confirmation.

File: CWE/Modules/General.cs [hunk 37]
```diff
@@ -234,5 +234,63 @@ public async Task CancelCampaign(SocketGuildUser user)
...
+        }
+
+        /// <summary>
+        /// Command used to transfer ownership of a tag.
+        /// </summary>
+        /// <param name="tagName">The name of the tag to transfer.</param>
+        /// <param name="newOwnerId">The new owner's ID value.</param>
+        /// <returns>The ownership action of transering a tag. <see cref="Task"/>.</returns>
+        [Command("tag transfer")]
+        [RequireTagAuthoriazation]
```
Comment (hendrikdemir) [inline CWE/Modules/General.cs]:
```suggestion
        [RequirePromoted]
```

File: CWE/Modules/General.cs (hunk 28 above)
Comment (hendrikdemir) [inline CWE/Modules/General.cs]:
```suggestion
        [RequirePromoted]
```

**Review by hendrikdemir — CHANGES_REQUESTED** [review]
Added comments regarding naming and i.e. data access layer.

**Comment by n-Ultima** [comment]
I went ahead and did all this in the most recent commit, along with editing the tag commands to match the naming changes in the DataAccessLayer.

File: CWE/Modules/General.cs (hunk 36 above)
Comment (n-Ultima) [reply CWE/Modules/General.cs]: This is already handles inside of the method, EditTagContent. It keeps the command modules looking cleaner in my opinion. If the executorId is not equal to the tag's current ownerId, then it throws an InvalidOperationException.

File: CWE.Data/DataAccessLayer.cs (hunk 5 above)
Comment (n-Ultima) [reply CWE.Data/DataAccessLayer.cs:151]: I've taken some thought into this. The way I see it, is that CWE already sends a message to the current channel if an exception is thrown. So, why not actually show the error here?

Comment (n-Ultima) [reply CWE.Data/DataAccessLayer.cs:151]: I could return a null, but then that would look ugly, so why not just throw a custom error?

**Comment by n-Ultima** [comment]
All the changes requested by Directoire have just been commited.

**Pull request merged**